datasync/downloader/queue.go

```diff
@@ -826,52 +831,54 @@ func (q *queue) deliver(id string, taskPool map[common.Hash]*types.Header, taskQ
 	var (
 		accepted int
 		failure  error
-		useful   bool
+		i        int
+		hashes   []common.Hash
 	)
-	for i, header := range request.Headers {
+	for _, header := range request.Headers {
 		// Short circuit assembly if no more fetch results are found
 		if i >= results {
 			break
 		}
-		// Reconstruct the next result if contents match up
-		index := int(header.Number.Int64() - int64(q.resultOffset))
-		if index >= len(q.resultCache) || index < 0 || q.resultCache[index] == nil {
-			failure = errInvalidChain
-			break
-		}
-		if err := reconstruct(header, i, q.resultCache[index]); err != nil {
+		// Validate the fields
+		if err := validate(i, header); err != nil {
 			failure = err
 			break
 		}
-		hash := header.Hash()
-
-		donePool[hash] = struct{}{}
-		q.resultCache[index].Pending--
-		useful = true
-		accepted++
+		hashes = append(hashes, header.Hash())
+		i++
+	}
 
+	for _, header := range request.Headers[:i] {
+		if res, stale, err := q.resultCache.GetDeliverySlot(header.Number.Uint64()); err == nil {
+			reconstruct(accepted, res)
+		} else {
+			// else: between here and above, some other peer filled this result,
+			// or it was indeed a no-op. This should not happen, but if it does it's
+			// not something to panic about
+			logger.Error("Delivery stale", "stale", stale, "number", header.Number.Uint64(), "err", err)
+			failure = errStaleDelivery
+		}
 		// Clean up a successful fetch
-		request.Headers[i] = nil
-		delete(taskPool, hash)
+		delete(taskPool, hashes[accepted])
+		accepted++
 	}
 	// Return all failed or missing fetches to the queue
 	for _, header := range request.Headers {
-		if header != nil {
-			taskQueue.Push(header, -int64(header.Number.Uint64()))
-		}
+		taskQueue.Push(header, -int64(header.Number.Uint64()))
 	}
 	// Wake up WaitResults
 	if accepted > 0 {
 		q.active.Signal()
 	}
+
 	// If none of the data was good, it's a stale delivery
 	if failure == nil {
 		return accepted, nil
 	}
 	if errors.Is(failure, errInvalidChain) {
```

Comment:
Not related to this change, however, `failure` can be written in the line 844 or 859.
As it is written as `errStableDelivery` at the line 859, only possibility of being written as `errInvalidChain` is by line 844. 

`deliver` is called from the two places `DeliverBodies` and `DeliverReceipts` and hence there are two kinds of `validate` function from both places. But both `validate` functions do not return `errInvalidChain` as you can check below. So maybe this line does not have any effect. Just FYI...

```
func (q *queue) DeliverBodies(id string, txLists [][]*types.Transaction) (int, error) {
	q.lock.Lock()
	defer q.lock.Unlock()

	validate := func(index int, header *types.Header) error {
		if types.DeriveSha(types.Transactions(txLists[index])) != header.TxHash {
			return errInvalidBody <<<<<<<<<
		}
		return nil <<<<<<<<<
	}

	reconstruct := func(index int, result *fetchResult) {
		result.Transactions = txLists[index]
		result.SetBodyDone()
	}
	return q.deliver(id, q.blockTaskPool, q.blockTaskQueue, q.blockPendPool, bodyReqTimer, len(txLists), validate, reconstruct)
}
```

```
func (q *queue) DeliverReceipts(id string, receiptList [][]*types.Receipt) (int, error) {
	q.lock.Lock()
	defer q.lock.Unlock()
	validate := func(index int, header *types.Header) error {
		if types.DeriveSha(types.Receipts(receiptList[index])) != header.ReceiptHash {
			return errInvalidReceipt <<<<<<<<<
		}
		return nil <<<<<<<<<
	}

	reconstruct := func(index int, result *fetchResult) {
		result.Receipts = receiptList[index]
		result.SetReceiptsDone()
	}
	return q.deliver(id, q.receiptTaskPool, q.receiptTaskQueue, q.receiptPendPool, receiptReqTimer, len(receiptList), validate, reconstruct)
}
```

Reply:
You are right. The invalid chain error was before this change in line 839. I think it is okay to remove it, but let me do this on another PR.

Reply:
@jeongkyun-oh If it also exists in go-ethereum, I think it is good for you to contribute your change to geth 😄